dpd-client/tests/integration_tests/common.rs

```diff
@@ -929,6 +929,153 @@ pub fn gen_udp_routed_pair(
+pub fn gen_geneve_packet(
+    src: Endpoint,
+    dst: Endpoint,
+    inner_type: u16,
+    vni: u32,
+    geneve_opts: &[OxideGeneveOption],
+    payload: &[u8],
+) -> Packet {
+    let udp_stack = match src.get_ip("src").unwrap() {
+        IpAddr::V4(_) => {
+            vec![inner_type, ipv4::IPPROTO_UDP.into(), eth::ETHER_IPV4]
+        }
+        IpAddr::V6(_) => {
+            vec![inner_type, ipv6::IPPROTO_UDP.into(), eth::ETHER_IPV6]
+        }
+    };
+
+    let mut pkt = Packet::gen(src, dst, udp_stack, Some(payload)).unwrap();
+    let geneve = pkt.hdrs.geneve_hdr.as_mut().unwrap();
+    geneve.vni = vni;
+
+    for opt in geneve_opts {
+        match opt {
+            OxideGeneveOption::External => {
+                #[rustfmt::skip]
+                geneve.options.extend_from_slice(&[
+                    // First 2 bytes: Geneve option class (0x0129)
+                    // The OXIDE vendor-specific class identifier
+                    0x01, 0x29,
+                    // Third byte: Critical bit (0) + Option type (1)
+                    0x00,
+                    // reserved + body len
+                    0x00,
+                ]);
+            }
+            OxideGeneveOption::Multicast(tag) if *tag < 3 => {
+                #[rustfmt::skip]
+                geneve.options.extend_from_slice(&[
+                    // First 2 bytes: Geneve option class (0x0129)
+                    // The OXIDE vendor-specific class identifier
+                    0x01, 0x29,
+                    // Third byte: Critical bit (0) + Option type (1)
+                    // Type 1 represents multicast tagged packets
+                    0x01,
+                    // Fourth byte: Option(s) length
+                    0x01,
+                    // Fifth byte: Tag value (encoded in the data)
+                    (tag & 0x03) << 6,
+                    // Sixth byte: reserved
+                    0x00,
+                    // Seventh byte
+                    0x00,
+                    // Eighth byte
+                    0x00,
+                ]);
+            }
+            OxideGeneveOption::Mss(mss) => {
+                #[rustfmt::skip]
+                geneve.options.extend_from_slice(&[
+                    // First 2 bytes: Geneve option class (0x0129)
+                    // The OXIDE vendor-specific class identifier
+                    0x01, 0x29,
+                    // Third byte: Critical bit (0) + Option type (1)
+                    0x02,
+                    // reserved + body len
+                    0x01,
+                ]);
+                geneve.options.extend_from_slice(&mss.to_be_bytes()[..]);
+            }
+            _ => {
+                panic!("illegal specification for option: {opt:?}")
+            }
+        }
+    }
+
+    let extra_bytes = geneve.options.len() as u16;
+    assert_eq!(
+        extra_bytes % 4,
+        0,
+        "Geneve options must be pushed in 4B blocks, found {extra_bytes}",
```

Comment:
Huh.  I didn't know `assert` let you provide an explanation.  That's very handy.